### ISSUE TITLE
File uploader cumulative size check 

### ISSUE THREAD

**Issue by rianfowler** [usds/us-forms-system#317]
**Is your feature request related to a problem? Please describe.**
File uploader should allow check on cumulative size of all uploaded files. We have a cumulative limit on one of the VA systems we're integrating with. 

**Describe the solution you'd like**
Should add an error when the cumulative file size exceeds set limit 

**Describe alternatives you've considered**
I couldn't find a callback I could pass in that has access to the file information and that can edit the errors object. AFAIK `ui:validations` only has access to the errors added by [uploadFile](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/actions.js#L165)

**Additional context**
 AFAIK this will require
- updating the signature of [this](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/actions.js#L165) function to include the size of all of the files or some interim data that can be derived from. 
- An additional uiOption check against the cumulative file size similar to [this](https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/actions.js#L167)

**Comment by dmethvin-gov** [comment]
This raises a bunch of UX issues about how a situation like that is treated. It's especially important with the VA-style forms where you can't proceed tot he next page until you've resolved an error.  It doesn't feel like it should be the job of the `uploadFile` function to track the cumulative size of all uploads on the form and throw an error for this particular field when it is the straw that breaks the camel's back.

It seems like most of the heavy lifting goes on in [`uiOptions.createPayload`]( https://github.com/usds/us-forms-system/blob/37c569781d188be421eed28e9b5987aeea3e6370/src/js/actions.js#L204) which is outside the library. Could that function track the current attachments and their cumulative size, and let the user know via some on-screen tally?

**Comment by rianfowler** [comment]
File size checks are already done in `uploadFile` so it makes sense to put all of the size checks in that on place and to manage it with the same configuration. Spreading that behavior around makes it harder to maintain. 

AFAIK `createPayload` doesn't have access to the error state. We could probably track the cumulative file size in a curried function but setting the error state would require reaching outside of the form system and plugging into the error state somehow.  

Semantically it doesn't make sense that `uploadFile` or `createPayload` do any checks. The checks should be in a validation but in this context, the validators only have access to the errors created by `uploadFile`

**Comment by dmethvin-gov** [comment]
At the moment the forms system design wants to attach errors to a specific form entry, and I agree you're staying to that pattern here. This is a *global* issue so I'm not sure that it makes sense to the user to view this as a problem with this particular form entry. Let's say there's a 10MB limit and they've already uploaded a big 9.8MB file. The tiny little 250KB file they upload next isn't really where the problem lies, it just pushed them over quota.

That's why I focused on UX. At some point, say on page 5 after uploading 3 other files on previous pages, the user will try to do a 4th file and it will give an error for the 4th file field something like, "Sorry, the 250KB file you just tried to upload exceeds the total of 10MB allowed." The next sentence should explain how they can fix the problem.

If this is a required field, they can't move forward. They can go back and remove one of the files on page 1 through 4 but they would need to find the files they attached and try to figure out whether there was a particular file that caused most of the overage. Then they would need to do the math in their head to see if they made it small enough to proceed.

So those are the UX concerns. Then there is the code side. To my mind the stuff below is all secondary though, we should determine how the user is notified and what we expect them to do to fix the problem before changing code.

Right now`uploadFile` isn't notified if an uploaded file is removed; it's just taken out of the form data by the FileField control that removed it. If we entrust `uploadFile` with the job of tracking global total bytes in order to enforce a quota, we'd need to notify it that files already uploaded need to be removed from the total upload count.

However, uploading happens asynchronously. Multiple uploads can be going on at once.  In the case there is some error returned by the server we would need to subtract the errored file size from the quota once the server returns the status. The same for the situation where the user cancels the upload. On any change to attachments we need to redo the global count, run through all the attachments on the form, and clear any "over quota" errors on any file upload field.

**Comment by rianfowler** [comment]
I was able to add a validator to the outer object that has access to the file names and the outer form object. Below I've thrown a test error on this for demoing purposes. This validator is fired when files are added or removed and it has access to the full list of file names / indexes so I think with the file size information captured in `createPayload` we have what we need to know when this is in an error state.

Couple of caveats / challenges:

- this error only shows up when the user clicks next- ideally this would allow the upload of the file but inform them that they've exceeded their limit (maybe we're okay with this limitation for now and this becomes a later enhancement) 
- the individual file sizes are not displayed so it's difficult to know which files are contributes what amount to the cumulative payload size (we know the file size before the upload but would only display it after a successful upload)

![screen shot 2018-11-13 at 14 32 18](https://user-images.githubusercontent.com/4998130/48444498-59d07100-e751-11e8-8e35-1b7240ab4cb2.png)

We have some copy suggestions in this [ticket](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/14562) but I think this might need some UX work. 

Other things that I need to work out based on your feedback:
- how do we handle file upload errors (e.g. does the validator fire and does it have enough info to detect this scenario)
- I don't completely understand the multiple page file upload scenario. Would this be a scenario where multiple pages have file upload UIs and there is a cumulative across all pages?

**Comment by dmethvin-gov** [comment]
Thanks for mocking this up, I think it helps to demonstrate the UX issues!

> I don't completely understand the multiple page file upload scenario. Would this be a scenario where multiple pages have file upload UIs and there is a cumulative across all pages?

Yeah, you could have different pages of your multi-page form that had file field attachments. 

AFAICT right now we require the upload to complete before you can press Next to go to the next page? I didn't experiment but the code seemed to say it was treating a file that is still uploading as an "upload in progress" error. That simplifies the cases that have to be handled because all the validation is contained to one page.

**Comment by dmethvin-gov** [comment]
Since this is rather specific to the VA's needs and requires UX work, I'll close this. @rianfowler feel free to port this ticket over the the VA repo and link back to it if it's still in the queue.
